fabric-item-api-v1/src/main/java/net/fabricmc/fabric/api/item/v1/CustomItemSettingType.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.item.v1;
+
+import java.util.function.Supplier;
+
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.NotNull;
+
+import net.minecraft.item.Item;
+
+import net.fabricmc.fabric.impl.item.FabricItemInternals;
+
+/**
+ * A type of setting that can be passed to item constructors.
+ * This feature can be used by mods to add non-standard settings
+ * to items in a way that is compatible with other mods that add
+ * settings to items.
+ *
+ * @param <T> the type of the setting to be attached
+ */
+public final class CustomItemSettingType<T> {
+	private final Supplier<@NotNull T> defaultValue;
+
+	private CustomItemSettingType(Supplier<@NotNull T> defaultValue) {
+		this.defaultValue = defaultValue;
+	}
+
+	// There are a lot of these for convenience sake.
+	public static <T> CustomItemSettingType<T> of(Supplier<T> defaultValue) {
```

Comment:
Imo the supplier version should have a different name from the direct getting version, and create another 
`public static <T> CustomItemSettingType<T> of(T defaultValue) {`

Reply:
The reason I use a supplier for non-primitive types is to ensure safety. Primitive types are passed by value, so it's safe to offer the same one to every item. Items might mutate the resulting value though, which would cause problems if we passed the same `ArrayList` for example to every item.

Reply:
I mean you can have a default value like a collection empty list. This isn't really a problem, as there are way more ways for bad code to break things, especially for minecraft where there's no strict scrutiny to prevent some other parts of code from intentionally breaking stuff.

Reply:
That's absolutely true. I still think it's worth providing that safety where we can.